aws-lc-rs/src/rsa/encryption.rs

```diff
@@ -0,0 +1,359 @@
+pub struct PrivateDecryptingKey(RsaEvpPkey);
+
+impl PrivateDecryptingKey {
+    fn new(key: LcPtr<EVP_PKEY>) -> Result<Self, Unspecified> {
+        Ok(Self(RsaEvpPkey::new(key, UsageContext::Decryption)?))
+    }
+
+    /// Generate a new RSA private key for use with asymmetrical encryption.
+    ///
+    /// # Errors
+    /// * `Unspecified` for any error that occurs during the generation of the RSA keypair.
+    pub fn generate(size: KeySize) -> Result<Self, Unspecified> {
+        let key = generate_rsa_key(size.bit_len(), false)?;
+        Self::new(key)
+    }
+
+    /// Generate a RSA `KeyPair` of the specified key-strength.
+    ///
+    /// Supports the following key sizes:
+    /// * `EncryptionKeySize::Rsa2048`
+    /// * `EncryptionKeySize::Rsa3072`
+    /// * `EncryptionKeySize::Rsa4096`
+    ///
+    /// # Errors
+    /// * `Unspecified`: Any key generation failure.
+    #[cfg(feature = "fips")]
+    pub fn generate_fips(size: KeySize) -> Result<Self, Unspecified> {
+        let key = generate_rsa_key(size.bit_len(), true)?;
+        Self::new(key)
+    }
+
+    /// Construct a `PrivateDecryptingKey` from the provided PKCS#8 (v1) document.
+    ///
+    /// Supports RSA key sizes between 2048 and 8192 (inclusive).
+    ///
+    /// # Errors
+    /// * `Unspecified` for any error that occurs during deserialization of this key from PKCS#8.
+    pub fn from_pkcs8(pkcs8: &[u8]) -> Result<Self, KeyRejected> {
+        let evp_pkey = LcPtr::try_from(pkcs8)?;
+        Self::new(evp_pkey).map_err(|_| KeyRejected::unexpected_error())
+    }
+
+    /// Returns a boolean indicator if this RSA key is an approved FIPS 140-3 key.
+    #[cfg(feature = "fips")]
+    #[allow(clippy::missing_panics_doc)]
+    #[must_use]
+    pub fn is_valid_fips_key(&self) -> bool {
+        self.0.is_valid_fips_key()
+    }
+
+    /// Returns the RSA key size in bytes.
+    #[must_use]
+    pub fn key_size(&self) -> usize {
+        self.0.key_size()
+    }
+
+    /// Retrieves the `PublicEncryptingKey` corresponding with this `PrivateDecryptingKey`.
+    ///
+    /// # Errors
+    /// * `Unspecified` for any error that occurs computing the public key.
+    pub fn public_key(&self) -> Result<PublicEncryptingKey, Unspecified> {
+        if 1 != unsafe { EVP_PKEY_up_ref(*self.0.key) } {
+            return Err(Unspecified);
+        };
+        Ok(PublicEncryptingKey(RsaEvpPkey::new(
+            LcPtr::new(*self.0.key)?,
+            UsageContext::Encryption,
+        )?))
+    }
+
+    /// Decrypts the contents in `ciphertext` and writes the corresponding plaintext to `output`.
+    ///
+    /// # Sizing `output`
+    /// For `OAEP_SHA1_MGF1SHA1`, `OAEP_SHA256_MGF1SHA256`, `OAEP_SHA384_MGF1SHA384`, `OAEP_SHA512_MGF1SHA512` a safe
+    /// `output` length is the key size in bytes. The RSA key size in bytes can be retrieved using [`Self::key_size`].
+    /// The actual output length will at most `rsaKeySizeBytes - (2 * hashLengthBytes) - 2`.
```

Comment:
There are two hashes involved, and currently they're the same. If the hashes were to differ, I believe it would be the output length of OAEP hash that determines the length here. Should we clarify: `rsaKeySizeBytes - (2 * oaepHashLengthBytes) - 2`?